syntax = "proto2";

package rwe.proto;

message LoadingStatusMessage
{
    enum Status
    {
        Loading = 0;
        Ready = 1;
    }

    required Status status = 1;
}

message SimVector
{
    required float x = 1;
    required float y = 2;
    required float z = 3;
}

message MoveOrder
{
    required SimVector destination = 1;
}

message AttackOrder
{
    oneof target
    {
        int32 unit = 1;
        SimVector ground = 2;
    }
}

message BuildOrder
{
    required string unit_type = 1;
    required SimVector position = 2;
}

message CompleteBuildOrder
{
    required int32 unit = 1;
}

message GuardOrder
{
    required int32 unit = 1;
}

message ReclaimOrder
{
    oneof target {
        int32 unit = 1;
        int32 feature = 2;
    }
}

message PlayerUnitCommand
{
    message IssueOrder
    {
        enum IssueKind
        {
            Immediate = 0;
            Queued = 1;
        }

        required IssueKind issue_kind = 1;
        oneof order
        {
            MoveOrder move = 2;
            AttackOrder attack = 3;
            BuildOrder build = 4;
            CompleteBuildOrder complete_build = 5;
            GuardOrder guard = 6;
            ReclaimOrder reclaim = 7;
        }
    }

    message ModifyBuildQueue
    {
        required int32 count = 1;
        required string unit_type = 2;
    }

    message Stop
    {
    }

    message SetFireOrders
    {
        enum FireOrders
        {
            HoldFire = 0;
            ReturnFire = 1;
            FireAtWill = 2;
        }

        required FireOrders orders = 1;
    }

    message SetOnOff
    {
        required bool isOn = 1;
    }

    required int32 unit = 1;

    oneof command {
        IssueOrder order = 2;
        Stop stop = 3;
        SetFireOrders set_fire_orders = 4;
        SetOnOff set_on_off = 5;
        ModifyBuildQueue modify_build_queue = 6;
    }
}

message PlayerPauseGameCommand
{
}

message PlayerUnpauseGameCommand
{
}

message PlayerCommand
{
    oneof command
    {
        PlayerUnitCommand unit_command = 1;
        PlayerPauseGameCommand pause = 2;
        PlayerUnpauseGameCommand unpause = 3;
    }
}

message GameUpdateMessage
{
    message PlayerCommandSet
    {
        repeated PlayerCommand command = 1;
    }
    required int32 next_command_set_to_receive = 1;
    required int32 next_command_set_to_send = 2;
    required int32 ack_delay = 3;
    required int32 current_scene_time = 4;
    repeated PlayerCommandSet command_set = 5;
    required uint32 player_id = 6;
    required int32 packet_id = 7;
    required int32 next_game_hash_to_send = 8;
    required int32 next_game_hash_to_receive = 9;
    repeated int32 game_hashes = 10;
}

message NetworkMessage
{
    oneof message
    {
        LoadingStatusMessage loading_status = 1;
        GameUpdateMessage game_update = 2;
    }
}
